drgn_tools/task.py

```diff
@@ -161,3 +170,91 @@ def task_cpu(task: drgn.Object) -> int:
     if has_member(task, "cpu"):
         return task.cpu.value_()
     return task_thread_info(task).cpu.value_()
+
+
+def format_nanosecond_duration(nanosecs: int) -> str:
+    """
+    :returns: conversion of nanoseconds to [dd hh:mm:ss.ms] format
+    """
+    secs = nanosecs_to_secs(nanosecs)
+    dd, rem = divmod(secs, 86400)
+    hh, rem = divmod(rem, 3600)
+    mm, secs = divmod(rem, 60)
+    return "%02ld %02ld:%02ld:%.3f" % (dd, hh, mm, secs)
+
+
+def get_pid(task: Object) -> int:
+    """
+    :returns: PID of the task
+    """
+    return task.pid.value_()
+
+
+def get_command(task: Object) -> str:
+    """
+    :returns: name of the command
+    """
+    return escape_ascii_string(task.comm.string_())
+
+
+def show_tasks_last_runtime(prog: Program, tasks: Iterable[Object]) -> None:
+    """
+    Display task information in their last arrival order.
+    """
+    print(
+        "LAST_ARRIVAL".ljust(25)
+        + "STATE".ljust(18)
+        + "PID".ljust(20)
+        + "TASK".ljust(20)
+        + "CPU".ljust(15)
+        + "COMMAND".ljust(20)
+    )
+    tasklist = []
+    for t in tasks:
+        cpu = task_cpu(t)
+        pid = get_pid(t)
+        state = task_state_to_char(t)
+        task = find_task(prog, pid)
+        command = get_command(t)
+        time_nanosec = task_lastrun2now(t)
+        last_arrival = format_nanosecond_duration(time_nanosec)
+
+        tasklist.append(
+            (
+                time_nanosec,
+                "%-25s %-15s %-20s %-20s %-10s %10s"
+                % (
+                    last_arrival,
+                    state,
+                    pid,
+                    hex(task.value_()),
+                    cpu,
+                    command,
+                ),
+            )
+        )
+    tasklist.sort()
+    for t in tasklist:
+        print(t[1])
```

Comment:
I really should have noticed this in the first round of review, but we have a function `print_table()` which is really great. I looked at the output from the test runs, and while your table is good, it is pretty wide. For example, the "STATE" field gets a lot of space, even though it only contains one character. Similar with PID and CPU, which are really wide. The benefit of `print_table()` is that it will determine appropriate column widths for you, so that the table will get automatically sized.

Here's a suggested rewrite which could use it:

```python
    rows = [["LAST_ARRIVAL", "ST", "PID", "TASK", "CPU", "COMMAND"]]
    tasks = list(tasks)
    tasks.sort(key=task_lastrun2now)
    for t in tasks:
        cpu = task_cpu(t)
        pid = get_pid(t)
        state = task_state_to_char(t)
        command = get_command(t)
        time_nanosec = task_lastrun2now(t)
        last_arrival = format_nanosecond_duration(time_nanosec)
        rows.append([last_arrival, state, pid, hex(t.value_()) cpu, command])
    print_table(rows)
```